File: structural/proxy/typescript/proxy.ts
import { IUserService } from './interface';
import { UserService } from './service';

export class UserServiceProxy implements IUserService {
	private realUserService: UserService;
	private cache: Map<number, string>;

	constructor(realUserService: UserService) {
		this.realUserService = realUserService;
		this.cache = new Map<number, string>();
	}

	async getUser(id: number): Promise<string> {
		if (this.cache.has(id)) {
			console.log(`Returning cached user with id ${id}...`);
			return Promise.resolve(this.cache.get(id) as string);
		}

		const user = await this.realUserService.getUser(id);
		this.cache.set(id, user);
		return user;
	}
}

File: structural/proxy/python/proxy.py
from interface import UserServiceModel
from service import UserService


class UserServiceProxy(UserServiceModel):
    def __init__(self, real_user_service: UserService):
        self.real_user_service = real_user_service
        self.cache = {}

    def get_user(self, user_id: int) -> str:
        if user_id in self.cache:
            print(f"Returning cached user with id {user_id}...")
            return self.cache[user_id]

        user = self.real_user_service.get_user(user_id)
        self.cache[user_id] = user
        return user
